internals/overlord/servstate/handlers.go

```diff
@@ -167,27 +180,34 @@ func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
 // and is running.
 //
 // It also returns a message to add to the task's log, or empty string if none.
-func (m *ServiceManager) serviceForStart(config *plan.Service) (service *serviceData, taskLog string) {
+func (m *ServiceManager) serviceForStart(config *plan.Service, workload *Workload) (service *serviceData, taskLog string) {
 	m.servicesLock.Lock()
 	defer m.servicesLock.Unlock()
 
+	var w *Workload
+	if workload != nil {
+		w = workload.copy()
+	}
+
 	service = m.services[config.Name]
 	if service == nil {
 		// Not already started, create a new service object.
 		service = &serviceData{
-			manager: m,
-			state:   stateInitial,
-			config:  config.Copy(),
-			logs:    servicelog.NewRingBuffer(maxLogBytes),
-			started: make(chan error, 1),
-			stopped: make(chan error, 2), // enough for killTimeElapsed to send, and exit if it happens after
+			manager:  m,
+			state:    stateInitial,
+			config:   config.Copy(),
+			workload: w,
+			logs:     servicelog.NewRingBuffer(maxLogBytes),
+			started:  make(chan error, 1),
+			stopped:  make(chan error, 2), // enough for killTimeElapsed to send, and exit if it happens after
 		}
 		m.services[config.Name] = service
 		return service, ""
 	}
 
 	// Ensure config is up-to-date from the plan whenever the user starts a service.
 	service.config = config.Copy()
+	service.workload = w
```

Comment:
Why not copy the line above ? :)

Reply:
Because `workload` here might be `nil`, and calling `workload.copy()` could potentially dereference this `nil` pointer. I made the code a bit more easy to follow though, thanks!